soot-infoflow-summaries/src/soot/jimple/infoflow/methodSummary/postProcessor/SummarySourceContextAndPath.java

```diff
@@ -344,20 +344,22 @@ private AccessPath matchAccessPath(AccessPath curAP, Value base, SootField field
 				final Collection<AccessPathFragment[]> bases = manager.getAccessPathFactory()
 						.getBaseForType(base.getType());
 				if (bases != null) {
-					for (AccessPathFragment[] xbase : bases) {
-						if (xbase[0].getField() == field) {
-							// Build the access path against which we have
-							// actually matched
-							AccessPathFragment[] cutFragments = new AccessPathFragment[curAP.getFragmentCount()
-									+ xbase.length];
-
-							System.arraycopy(xbase, 0, cutFragments, 0, xbase.length);
-							System.arraycopy(curAP.getFragments(), 0, cutFragments, xbase.length,
-									curAP.getFragmentCount());
-
-							return manager.getAccessPathFactory().createAccessPath(curAP.getPlainValue(),
-									curAP.getBaseType(), cutFragments, curAP.getTaintSubFields(), false, false,
-									curAP.getArrayTaintType());
+					synchronized (bases) {
```

Comment:
Ich verstehe den Sinn hinter diesem Ansatz nicht. Dass du einen Monitor auf dem Set hälst, hindert niemanden daran, parallel das Set zu verändern. Du müsstest auch das Reinschreiben in das Set synchronisieren auf demselben Monitor-Objekt (dem Set), damit das etwas bringt.

Reply:
Critical Sections verhindern lediglich, dass zwei Threads gleichzeitig eine CS durchlaufen, die durch dasselbe Monitor Objekt geschützt ist. Mehr macht eine CS nicht.

Reply:
ok, das ``ConcurrentSet`` lockt auf demselben Objekt, dann passt das glücklicherweise.